### PR TITLE
DrivAerML: 4L/256d throughput — more epochs in wall-clock

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Reducing hidden dim from 320→256 and heads from 5→4 (keeping 64 per head) makes each epoch ~40% faster, allowing the model to see substantially more gradient steps in the 180-min wall-clock budget. If the 4L/320d result at epoch 257 is still on an improving trajectory, a narrower but faster model may reach a better minima by completing ~360 epochs in the same wall-clock time. This tests whether throughput beats raw capacity at this scale.

## Instructions
Run from `target/icml2026` with the following command (copy exactly):

```bash
SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group drivaerml-width-scaling
```

Changes from baseline: `--model-hidden-dim 256` (was 320), `--model-heads 4` (was 5). All other hyperparameters unchanged — this isolates the width/throughput tradeoff.

## Baseline
Current DrivAerML best (PR #2648, W&B run `qx7z7if3`):
- `val_primary/surface_rel_l2_pct`: **5.027%** (val) / 6.244% (test)
- Epoch: 257 of 180-min budget
- External competition target: **< 3.71%**

Baseline reproduce command:
```bash
cd target/icml2026 && SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 320 --model-heads 5 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```

Beat `val_primary/surface_rel_l2_pct < 5.027%` to be a winner. Also report final epoch count so we can measure actual throughput gain.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2759 because ADVISOR: Closing — 4L/256d architecture is two generations behind. Baseline moved from 4L/320d to 4L/512d (4.619%). The throughput advantage of 4L/256d no longer compensates for the capacity gap at 180-min budgets..

**Comment by tcapelle** [comment]
STUDENT zenitsu:

## Results

### DrivAerML 4L/256d/4H throughput test (lr=5e-4)

| Metric | Baseline (PR #2648, 4L/320d) | This Run (4L/256d) | Delta |
|--------|------------------------------|---------------------|-------|
| **Best val_primary/surface_rel_l2_pct** | **5.027%** (ep257) | **5.24%** (ep262) | **+0.21%** ❌ |
| test_primary/surface_rel_l2_pct | 6.244% | 6.40% | +0.16% |
| Final training loss | — | 0.0066 | — |
| Epochs completed | 257 | 266 | +9 (+3.5%) |

### Verdict: Throughput advantage insufficient

The 4L/256d model did **not** beat the 5.027% baseline. Best val was 5.24% — 0.21% short. The hypothesis that throughput (more epochs) compensates for reduced capacity did not hold in this test.

### Key Observations

**Throughput gain was minimal**: 4L/256d completed 266 epochs vs 4L/320d's 257 — only 3.5% more epochs, far below the hypothesized ~40% throughput advantage. The per-epoch time difference was small, suggesting the bottleneck is data loading/eval, not model compute.

**Capacity matters more than throughput at this scale**: Despite more epochs, the narrower model consistently tracked ~1.5-2% behind the wider model at matched epoch counts:
- Ep50: 4L/256d 12.65% vs 4L/384d 10.56%
- Ep100: 4L/256d 8.42% vs 4L/384d 6.83%
- Ep200: 4L/256d 5.99% vs 4L/384d 5.11% (from PR #2694)

**Late-run improvement pattern**: The model was still setting new bests at ep262 (5.24%), suggesting it hasn't fully converged. With more wall-clock time it could potentially reach the baseline, but the convergence rate was decelerating.

**Wider val oscillations**: 4L/256d showed larger restart bumps (up to ~4% swings) compared to 4L/384d (~2% swings), suggesting the narrower model is more sensitive to LR restarts.

### Throughput Comparison
| Model | Epochs | Best Val | Time/Epoch | Val at ep257 |
|-------|--------|----------|------------|--------------|
| 4L/320d | 257 | 5.027% | ~0.70 min | 5.027% |
| 4L/256d | 266 | 5.24% | ~0.68 min | ~5.40% |

### Command Used
```bash
cd /workspace/senpai/target/icml2026 && SENPAI_TIMEOUT_MINUTES=180 SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --agent zenitsu --wandb-name "zenitsu/drivaerml-4L256d-throughput" \
  --wandb-group "drivaerml-width-scaling"
```

- **W&B run**: `tn8d9v06` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/tn8d9v06
- **Peak GPU memory**: ~1.2GB VRAM (batch-size 1, 50k surface points)

### Suggested Follow-ups

1. **Throughput hypothesis is weak for DrivAerML**: At batch-size 1 with 50k surface points, data loading dominates — model size has minimal impact on per-epoch time. This invalidates the core hypothesis
2. **Width scaling confirms 320d > 256d**: The 4L/320d baseline is superior. Consider testing 4L/352d or 4L/384d with lr=5e-4 and the longer timeout to push further
3. **Higher LR for 4L/256d**: lr=7e-4 dramatically helped 4L/384d (PR #2694, 4.74% val). It might help close the gap for 4L/256d too
4. **Model is still improving**: 5.24% at ep262 with clear downward trend — a longer budget might eventually match baseline, but 4L/320d reaches it faster
